### PR TITLE
Fixes 3877: Add required fields for API spec requests

### PR DESCRIPTION
**Pull request by Andrewgdewar**
## Summary

- Add required fields for API spec requests.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-3877

**Comment by swadeley** [comment]
Hi

How to test these changes?

If I take the first one:
```
   "definitions": {
        "api.AddUploadsRequest": {
            "type": "object",
            "required": [
                "artifacts",
                "uploads"
            ],
```
Looking at "add_upload" in ephemeral I do not see anything difenret to that in stage:
```
app.content_sources.rest_client.repositories_api.add_upload.
                                                                      allowed_values  call_with_http_info() headers_map           params_map           
                                                                      api_client            callable()            location_map          settings             
                                                                      attribute_map         collection_format_map openapi_types         validations        
```

**Comment by jlsherrill** [comment]
I don't know that the changes need to be tested, but if you regenerate the api bindings you will see that code will likely need to be changed.  for example, this method signature changes:

```
@@ -102,9 +102,12 @@ class ApiContentUnitSearchRequest(ModelNormal):
     ])

     @convert_js_args_to_python_args
-    def __init__(self, *args, **kwargs):  # noqa: E501
+    def __init__(self, search, *args, **kwargs):  # noqa: E501
         """ApiContentUnitSearchRequest - a model defined in OpenAPI

+        Args:
+            search (str): Search string to search content unit names
+
```

So  if you are using the generated bindings its clearer that 'search' is required to be specified.   This doesn't actually change the behavior of the API at all (hence why the tests all pass without generating the bindings)

File: pkg/api/content_units.go [hunk 1]
```diff
@@ -1,16 +1,16 @@
 package api
 
 type ContentUnitListRequest struct {
-	UUID   string `param:"uuid"`    // Identifier of the repository
-	Search string `query:"search"`  // Search string based query to optionally filter-on
-	SortBy string `query:"sort_by"` // SortBy sets the sort order of the result
+	UUID   string `param:"uuid" validate:"required"`    // Identifier of the repository
+	Search string `query:"search" validate:"required"`  // Search string based query to optionally filter-on
+	SortBy string `query:"sort_by" validate:"required"` // SortBy sets the sort order of the result
```
Comment (rverdile) [inline pkg/api/content_units.go:6]: search and sort_by are likely not required inputs

File: pkg/api/content_units.go [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
 type ContentUnitSearchRequest struct {
-	URLs   []string `json:"urls,omitempty"`  // URLs of repositories to search
-	UUIDs  []string `json:"uuids,omitempty"` // List of repository UUIDs to search
-	Search string   `json:"search"`          // Search string to search content unit names
-	Limit  *int     `json:"limit,omitempty"` // Maximum number of records to return for the search
+	URLs   []string `json:"urls,omitempty"`             // URLs of repositories to search
+	UUIDs  []string `json:"uuids,omitempty"`            // List of repository UUIDs to search
+	Search string   `json:"search" validate:"required"` // Search string to search content unit names
+	Limit  *int     `json:"limit,omitempty"`            // Maximum number of records to return for the search
 }
```
Comment (rverdile) [inline pkg/api/content_units.go:14]: The required inputs for ContentUnitSearchRequest are URLs and UUIDs. With that said, only one of them is required. So I'm not sure if it makes sense to mark them both as required, or neither?

In theory we allow an empty search, so I wouldn't say search is required

File: pkg/api/rpms.go
Comment (rverdile) [inline pkg/api/rpms.go:1]: For all of these content unit search and list requests, I believe only the UUID is required. 

For the ones where you can specify UUIDs or URLs, similar to above to the question above, not sure if it should require both or neither.

File: pkg/api/templates.go [hunk 3]
```diff
@@ -10,15 +10,15 @@ import (
...
+	Name            *string        `json:"name" validate:"required"`                                            // Name of the template
+	Description     *string        `json:"description"`                                                         // Description of the template
+	RepositoryUUIDS []string       `json:"repository_uuids" validate:"required"`                                // Repositories to add to the template
+	Arch            *string        `json:"arch" validate:"required"`                                            // Architecture of the template
+	Version         *string        `json:"version" validate:"required"`                                         // Version of the template
+	Date            *EmptiableDate `json:"date"`                                                                // Latest date to include snapshots for
+	OrgID           *string        `json:"org_id" readonly:"true" swaggerignore:"true" validate:"required"`     // Organization ID of the owner
+	User            *string        `json:"created_by" readonly:"true" swaggerignore:"true" validate:"required"` // User creating the template
```
Comment (rverdile) [inline pkg/api/templates.go:20]: These come from the header, but they're not required in the request body, so I wouldn't include "required" here

File: pkg/api/uploads.go [hunk 4]
```diff
@@ -3,8 +3,8 @@ package api
 import "time"
 
 type AddUploadsRequest struct {
-	Uploads   []Upload   `json:"uploads"`   // List of unfinished uploads
-	Artifacts []Artifact `json:"artifacts"` // List of created artifacts
+	Uploads   []Upload   `json:"uploads" validate:"required"`   // List of unfinished uploads
+	Artifacts []Artifact `json:"artifacts" validate:"required"` // List of created artifacts
```
Comment (xbhouse) [inline pkg/api/uploads.go:7]: i believe artifacts also isn't required, just the uploads. there isn't a public api to finish the uploads (which creates the artifacts) so someone using only the public apis would only ever send upload uuids

**Comment by Andrewgdewar** [comment]
Thanks for the review @rverdile @xbhouse!

**Comment by rverdile** [comment]
there's some tests failing :)

File: pkg/api/templates.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ type TemplateResponse struct {
 
 // We use a separate struct because version and arch cannot be updated
 type TemplateUpdateRequest struct {
-	UUID            *string        `json:"uuid" readonly:"true" swaggerignore:"true"`
+	UUID            *string        `json:"uuid" readonly:"true" swaggerignore:"true" validate:"required"`
```
Comment (rverdile) [inline pkg/api/templates.go:48]: When updating the template, the uuid comes from the url param and not from the request body, so I wouldn't mark it as required here

**Review by rverdile — COMMENTED** [review]
I thought I saw them before, but TemplateRequest and RepositoryRequest are missing

File: pkg/api/repositories.go [hunk 6]
```diff
@@ -146,11 +146,11 @@ func (r *RepositoryRequest) FillDefaults() {
 }
 
 type RepositoryIntrospectRequest struct {
-	ResetCount bool `json:"reset_count"` // Reset the failed introspections count
+	ResetCount bool `json:"reset_count" validate:"required"` // Reset the failed introspections count
```
Comment (rverdile) [inline pkg/api/repositories.go:149]: ResetCount is not required for RepositoryIntrospectRequest

**Comment by Andrewgdewar** [comment]
> I thought I saw them before, but TemplateRequest and RepositoryRequest are missing

Thanks for the feedback here, I did add to the template and repository request. Please confirm if those changes make sense.

File: pkg/api/content_units.go [hunk 7]
```diff
@@ -1,9 +1,9 @@
 package api
 
 type ContentUnitListRequest struct {
-	UUID   string `param:"uuid"`    // Identifier of the repository
-	Search string `query:"search"`  // Search string based query to optionally filter-on
-	SortBy string `query:"sort_by"` // SortBy sets the sort order of the result
+	UUID   string `param:"uuid" validate:"required"`   // Identifier of the repository
+	Search string `query:"search" validate:"required"` // Search string based query to optionally filter-on
```
Comment (rverdile) [inline pkg/api/content_units.go:5]: search nor UUID is required here. Search is an optional filter and the UUID comes from the URL, so it's not actually passed by the user through the request body.

**Comment by rverdile** [comment]
that's my last comment :)

**Comment by Andrewgdewar** [comment]
@rverdile @xbhouse Thank you for your thorough review here, I'm still learning 🍼 👶, I appreciate your patience!!

**Review by rverdile — APPROVED** [review]
lgtm!

**Comment by swadeley** [comment]
Hi

I found that I had to add `origin="external` to all repo create calls.
No changes to anything related to search this time (that had `origin="external`  from a while  back)

**Comment by rverdile** [comment]
Ah, it seems origin is actually not a required field. We should remove that then.

**Pull request merged**